std/string.d

```diff
@@ -221,10 +221,44 @@ inout(char)[] fromStringz(inout(char)* cString) @nogc @system pure nothrow {
 ///
 @system pure unittest
 {
-    assert(fromStringz(null) == null);
+    char* foo = null;
+    assert(fromStringz(foo) == null);
     assert(fromStringz("foo") == "foo");
 }
 
+/++
+    Params:
+        cString = A null-terminated c-style 16bit wide string.
+
+    Returns: A D-style array of `wchar` referencing the same string.  The
+    returned array will retain the same type qualifiers as the input.
+
+    $(RED Important Note:) The returned array is a slice of the original buffer.
+    The original data is not changed and not copied.
++/
+
+inout(wchar)[] fromStringz(inout(wchar)* cString) @nogc @system pure nothrow
+{
+    static size_t strlen16(const wchar* s)
```

Comment:
Some libraries have optimized architecture-specific versions of `wcslen`. You could use that when C's `wchar_t` defined in `core.stdc.stddef` is the same width as D's `wchar`.

Reply:
Please read https://github.com/dlang/phobos/pull/6438#discussion_r180185217 , unless I have misunderstood, using wcslen as I had coded originally was dismissed by @wilzbach in favour of consistency.

Reply:
I agree that we don't want `version (Windows)`, but in my opinion it would be perfectly safe to write `static if (wchar_t.sizeof == wchar.sizeof)`. @wilzbach what do you say?

Reply:
_Optimization is the root of all evil_ :P

I just think that it's really hard to optimize such a three-line loop and that we should value correctness over potential weird corner cases.
So before we go down the route of calling a runtime function that might behave differently, we should be sure it's worth it.
That's just my opinion though.

Reply:
(OTOH of course the case could be made that we should be able to trust the C runtime)

Reply:
@wilzbach Would it be worth adding another test this new function, for example, I wanted to be sure that it works correctly when surrogate pairs are used. I think my assumption is correct that surrogate pairs are always non-zero.

Reply:
A surrogate pair consists of a high surrogate in the half-open range `D800 .. DC00` followed by a low surrogate in the half-open range `DC00 .. E000`. So yes, surrogate pairs are always non-zero.

Reply:
Regarding how an optimized implementation can be faster than a naïve implementation at the expense of clarity and brevity, some read until the pointer is properly aligned then switch to full-word reads or vector operations.

Reply:
@GilesBathgate FWIW adding more tests is almost never a bad idea ;-)
@n8sh understood. I was just stating my opinion. I don't feel strongly about this and wouldn't object if Giles prefers to go with your proposal.

Reply:
@wilzbach Likewise!